### DISCUSSION TITLE
Highlight cursor line in active window; unhighlight cursor line in non-active windows

### DISCUSSION THREAD

**Discussion by armyers** [LazyVim/LazyVim#2031] [Q&A] [answered]
Hi,

I have two autocmds that will highlight the cursor line in the active window (the one that has the focus), and will deactive the cursor line highlights in any other visible windows that do not have the focus. I've put the following code in configs/autocmd.lua:

```
local autocmd = vim.api.nvim_create_autocmd
autocmd({ "FocusGained", "TabEnter", "WinEnter", "BufEnter", "BufReadPre", "BufRead", "BufWinEnter" }, {
  callback = function()
    vim.o.cursorline = true
  end,
})
autocmd({ "FocusLost", "TabLeave", "WinLeave", "BufLeave", "BufWinLeave" }, {
  callback = function()
    vim.o.cursorline = false
  end,
})
```

this has worked great (not perfect) when using a few other packaged nvim configs but it's not working so well with LazyVim. The problem is that when a popup notification happens, that is a new buffer/window, so there is a BufWinEnter event for the popup, but I don't see any kind of 'leave' event on the main editing buffer/window. And when the popup goes away, I do see a BufWinLeave event for the popup, but no 'Enter' event for the main editing buffer/window.   So the cursorline highlight goes away but doesn't come back.  

This is happening with the notification popups, the command line popup, and even the lines at the top of the buffer that show the current context.

Anyone have any ideas on how to make this work with LazyVim?

Thank you!

**Comment by jyuan0** [comment, marked as answer]
Here's what I've been using (copied from folke's config):
```lua
local cl_var = "auto_cursorline"
vim.api.nvim_create_autocmd({ "InsertLeave", "WinEnter" }, {
  group = vim.api.nvim_create_augroup("enable_auto_cursorline"),
  callback = function()
    local ok, cl = pcall(vim.api.nvim_win_get_var, 0, cl_var)
    if ok and cl then
      vim.api.nvim_win_del_var(0, cl_var)
      vim.wo.cursorline = true
    end
  end,
})
vim.api.nvim_create_autocmd({ "InsertEnter", "WinLeave" }, {
  group = vim.api.nvim_create_augroup("disable_auto_cursorline"),
  callback = function()
    local cl = vim.wo.cursorline
    if cl then
      vim.api.nvim_win_set_var(0, cl_var, cl)
      vim.wo.cursorline = false
    end
  end,
})
```

It's not the same events as you have, but hope it helps.

**Reply by armyers** [reply]
awesome! that's exactly what I needed. Here's what I ended up with:
```lua
local cl_var = "auto_cursorline"

vim.api.nvim_create_autocmd({ "WinEnter", "FocusGained" }, {
  group = vim.api.nvim_create_augroup("enable_auto_cursorline", { clear = true }),
  callback = function()
    local ok, cl = pcall(vim.api.nvim_win_get_var, 0, cl_var)
    if ok and cl then
      vim.api.nvim_win_del_var(0, cl_var)
      vim.o.cursorline = true
    end
  end,
})

vim.api.nvim_create_autocmd({ "WinLeave", "FocusLost" }, {
  group = vim.api.nvim_create_augroup("disable_auto_cursorline", { clear = true }),
  callback = function()
    local cl = vim.o.cursorline
    if cl then
      vim.api.nvim_win_set_var(0, cl_var, cl)
      vim.o.cursorline = false
    end
  end,
})
```

I determeined that  `BugWinEnter` and `BufWinLeave` were the problem events.

Also, I had to add `clear = true` on the augroup call to make it work. And I played with `vim.wo.cursorline` vs `vim.o.cursorline` and they both seem to behave the same, but I'm probably not using nvim in a way that would make a difference. Same thing for the `InsertEnter` and `InsertLeave` events; I'm probably not seeing a need for those because I'm not using nvim in the same way, or something.

One more question though....

When I startup nvim with the `-O` option and multiple files, such as
```bash
nvim -O autocmds.lua lazy.lua
```
I get highligted, underlined cursor lines in both windows. How would I change this so that the cursor line highlight/underline effects are only on the first window that has the focus?

Thank you again!

**Reply by jyuan0** [reply]
> How would I change this so that the cursor line highlight/underline effects are only on the first window that has the focus?

Sorry, don't know. I don't get cursor line highlighting in both windows if I open multiple files with `nvim -O`. Maybe it's the events? I'm also using nightly.

**Reply by armyers** [reply]
np. I'll play with the events and keep trying if it bothers me too much. :)

Thank you again for the help!
